### DISCUSSION TITLE
Hessian in CodeGenerator

### DISCUSSION THREAD

**Discussion by itbellix** [casadi/casadi#3436] [Q&A] [answered]
Hi.
I am using c-code generation from Python to export a CasADi function. I am exporting the information regarding the jacobian as follows:

```python
import casadi as ca

...

cg = ca.CodeGenerator('foo_jac')
F = ca.Function('F',[x],[y])
cg.add(F)
cg.add(F.jacobian())
cg.generate()
```

Is there a way to generate also the hessian of the function directly?

Since I want to use the generated function in a NLP (representing the equations of motion of my system), I am forced to solve the problem with the option `'hessian_approximation = 'limited-memory'` (in `ipopt`). However, when I use the code generator, I could also provide the analytical formulation of the hessian, and avoid this workaround, but I cannot find how to do this properly.

Thank you,
Italo

**Comment by jaeandersson** [comment]
There are a couple of different ways you can do it. The easiest is probably to start with creating an IPOPT instance in Python:
```python
import casadi as ca
x = ca.SX.sym('x')
nlp = dict(x = x, f = x*x)
solver = ca.nlpsol('solver', 'ipopt', nlp)
```
You can now code the NLP description as well as the functions that IPOPT needs to solve the problem, e.g. as follows:
```python
# Save to disk
cg = ca.CodeGenerator('foo')
cg.add(solver.oracle())  # Add NLP description 
all_cache_names = []
for name, fcn in solver.cache().items():
  all_cache_names.append(name)
  cg.add(fcn)  # Add solver specific functions
cg.generate()
print(f'Code generated solver specific functions: {all_cache_names}')
```
All the functions are now in a single C file, which can be compiled on some system:
```python
import os
os.system('gcc -shared foo.c -o foo.so')
```
On the other system, we can recreate the functions as follows:
```python
# Load code generated functions
oracle = ca.external('nlp', 'foo.so')
# Recreate cache
cache = dict()
for name in ['nlp_f', 'nlp_g', 'nlp_grad', 'nlp_grad_f', 'nlp_hess_l', 'nlp_jac_g']:
  cache[name] = ca.external(name, 'foo.so')
```
The oracle function and the cache dict should be enough to create IPOPT instances. For some reason, it looks like the constructor that takes oracle functions hasn't been exposed to the public API, so we need to make a symbolic description of the NLP:
```python
# Workaround: Looks like it's not possible to pass oracle instead of a symbolic NLP
oracle_in = oracle.mx_in()
oracle_out = oracle.call(oracle_in)
oracle_wrap = dict(zip(oracle.name_in() + oracle.name_out(), oracle_in + oracle_out))
```
We are now ready to create NLP instances:
```python
# Create IPOPT instance
solver2 = ca.nlpsol('solver2', 'ipopt', oracle_wrap, dict(cache = cache))
```
By passing `cache` above, it instructs the NLP interface to use the provided functions instead of the auto-generated ones.

**Comment by itbellix** [comment]
Dear @jaeandersson, thank you for your answer.
However, this is seems to be a procedure that is applicable to the whole NLP instance. What I am actually looking for is something (potentially) different, as basically I am creating my function `F` such that it captures the dynamics of my system, and then I am loading the corresponding library as an external function when I build and solve my NLP later. My hope is to save not only the jacobian, but also the hessian of `F` when generating the library, so that my NLP would be able to exploit that information too.

Does this make sense?

**Reply by jaeandersson** [reply, marked as answer]
The approach I outlined above includes the Hessian of the Lagrangian function (`nlp_hess_l`). But it assumes that you generate code for the NLP formulation specifically. If you want to generate code for a function `F` specifically, you can do that too, but it can be a bit messy if you want to capture all relevant derivative information. A minimal approach could be to include a reverse directional derivative and it's Jacobian:
```
cg = ca.CodeGenerator('foo_jac')
F = ca.Function('F',[x],[y])
cg.add(F)
cg.add(F.jacobian())
adj_F = F.reverse(1)
cg.add(adj_F)
cg.add(adj_F.jacobian())
cg.generate()
```
Generating code this way might be more versatile - you can then use `F` in different ways -  but it might also be a bit less efficient, but it might not matter much. You should get second derivatives and sparsities this way though.

**Reply by jaeandersson** [reply]
Also don't forget that there is the option of serializing `F` (i.e. saving it to disk symbolically) instead of generating C code for it. If you save it symbolically, no derivative information is lost.

**Reply by itbellix** [reply]
@jaeandersson , thank you for your quick reply.
I was indeed forgetting about the possibility of serializing my function. I will look into it, thanks again for your time!
